File: 2111-70-1910-remove-all-occurrences-of-a-substring/2111-70-1910-remove-all-occurrences-of-a-substring.py
class Solution:
    def removeOccurrences(self, s: str, part: str) -> str:
        def kmp(part):
            lps = [0] * len(part)
            j = 0
            for i in range(1, len(part)):
                while j and part[i] != part[j]:
                    j = lps[j - 1]
                if part[i] == part[j]:
                    j += 1
                    lps[i] = j
            return lps

        lps = kmp(part)
        sz = len(part)
        ans, stk = [], []
        j = 0

        for ch in s:
            ans.append(ch)
            stk.append(j)

            while j and ch != part[j]:
                j = lps[j - 1]

            if ch == part[j]:
                j += 1

            stk[-1] = j

            if j == sz:
                ans = ans[:-sz]
                stk = stk[:-sz]
                j = stk[-1] if stk else 0

        return "".join(ans)

File: 2111-70-1910-remove-all-occurrences-of-a-substring/2111-70-1910-remove-all-occurrences-of-a-substring.js
var removeOccurrences = function(s, part) {
    const kmp = (part) => {
        let lps = Array(part.length).fill(0);
        let j = 0;
        for (let i = 1; i < part.length; i++) {
            while (j && part[i] !== part[j]) j = lps[j - 1];
            if (part[i] === part[j]) lps[i] = ++j;
        }
        return lps;
    };

    let lps = kmp(part);
    let sz = part.length;
    let ans = [];
    let stk = [];
    let j = 0;

    for (let ch of s) {
        ans.push(ch);
        stk.push(j);

        while (j && ch !== part[j]) j = lps[j - 1];

        if (ch === part[j]) j++;

        stk[stk.length - 1] = j;

        if (j === sz) {
            ans.length -= sz;
            stk.length -= sz;
            j = stk.length ? stk[stk.length - 1] : 0;
        }
    }

    return ans.join('');
};
